File: python/src/wdl_model/model/expressions/wdl_index_access_operation.py
from __future__ import annotations

from dataclasses import dataclass

from .wdl_expression import ComponentType, WdlExpression


@dataclass
class WdlIndexAccessOperation(WdlExpression):
    target: WdlExpression | None = None
    index: WdlExpression | None = None

    def __str__(self) -> str:
        return f"{self.target or ''}[{self.index or ''}]"

    def componentType(self) -> ComponentType:
        return ComponentType.INDEX_ACCESS

File: typescript/src/model/expressions/wdl-index-access-operation.ts
/** Index access expression nodes. */
import { WdlExpressionComponentType, type WdlExpression } from './wdl-expression.js';

/** Models index lookup such as `arr[0]` or `map[key]`. */
export class WdlIndexAccessOperation implements WdlExpression {
  /** Creates an index access from the target expression and index expression. */
  public constructor(
    private targetValue?: WdlExpression,
    private indexValue?: WdlExpression,
  ) {}

  /** Returns the indexed target expression. */
  public getTarget(): WdlExpression | undefined {
    return this.targetValue;
  }
  /** Sets the indexed target expression. */
  public setTarget(target: WdlExpression | undefined): void {
    this.targetValue = target;
  }
  /** Returns the index expression. */
  public getIndex(): WdlExpression | undefined {
    return this.indexValue;
  }
  /** Sets the index expression. */
  public setIndex(index: WdlExpression | undefined): void {
    this.indexValue = index;
  }
  /** Returns the broad expression family for this node. */
  public componentType(): WdlExpressionComponentType {
    return WdlExpressionComponentType.IDX_OP;
  }
}
